### PR TITLE
Add Jinja Chat Template and Span Masking

### PR DESCRIPTION
**Pull request by fabianlim**
This is the companion PR to https://github.com/garrett361/open-instruct/pull/14 and https://github.com/garrett361/open-instruct/pull/13

To use the `span masking` encoding, must provide `--dataset_transform_fn sft_span_seach_mask_out sft_tulu_filter_v1`, as follows

```
python -m open_instruct.finetune \
    --tokenizer_name_or_path ... \
    --chat_template_name /path/to/chat/template.jinja \
    --max_seq_length 4096 \
    --dataset_mixer_list \
		.... 1.0 \
    --dataset_transform_fn sft_span_seach_mask_out sft_tulu_filter_v1 \
    --dataset_skip_cache False \
    --dataset_cache_mode local \
    --cache_dataset_only True \
    --dataset_local_cache_dir datasets/cache \
    --push_to_hub False \
    --try_launch_beaker_eval_jobs False

```

### CODE REVIEW

File: open_instruct/dataset_transformation.py [hunk 1]
```diff
@@ -826,6 +838,79 @@ def sft_tulu_tokenize_and_truncate_v1(row: Dict[str, Any], tokenizer: PreTrained
     return row
 
 
+def sft_span_seach_mask_out(row: Dict[str, Any], tokenizer: PreTrainedTokenizer, max_seq_length: int):
+
```
Comment (Swanand-Kadhe) [inline open_instruct/dataset_transformation.py:842]: I would suggest to add a docstring adapting from that of the companion function `sft_tulu_tokenize_and_truncate_v1`.

>"""
    This function encodes a single example into a format that can be used for sft training.
    Here, we assume each example has a 'messages' field. Each message in it is a dict with 'role' and 'content' fields.
    We use the `apply_chat_template` function from the tokenizer to tokenize the messages and prepare the input and label tensors.
    """

A suggestion:

> This function encodes a single example into a format that can be used for sft training (similar to `sft_tulu_tokenize_and_truncate_v1`). Instead of performing label masking iteratively, this function performs masking via span search and can handle complex chat templates with thinking.

**Review by Swanand-Kadhe — APPROVED** [review]
This looks good to me. I just have one suggestion. 
Many thanks.

File: open_instruct/dataset_transformation.py [hunk 2]
```diff
@@ -826,6 +838,79 @@ def sft_tulu_tokenize_and_truncate_v1(row: Dict[str, Any], tokenizer: PreTrained
...
+        end_tag = torch.tensor([end_tag])
+
+        # - prep
+        labels = input_ids.clone()
+
+        # - lengths
+        k, n = asst_tag.shape[1], labels.shape[1]
+        k1 = end_tag.shape[1]
+        
+        if n >= max(k, k1):
```
Comment (garrett361) [inline open_instruct/dataset_transformation.py:871]: A little confused about when `n >= max(k, k1)` is false. When?

**Review by garrett361 — APPROVED** [review]
Left several comments. Believe I followed the logic, but could you please add a small test case, too?  Happy to merge after comments are addressed and maybe a test.

File: open_instruct/dataset_transformation.py [hunk 3]
```diff
@@ -826,6 +838,79 @@ def sft_tulu_tokenize_and_truncate_v1(row: Dict[str, Any], tokenizer: PreTrained
...
+    # Span label masking strategy
+    # - search spans asst_tag ... end_tag
+    # - all such spans are left unmasked
+    # - if an asst_tage is undetected due to tokenization issues
+    #   then a span can be erronously masked
+    # - to avoid this, use tags that are guarded by special tokens
+    def masking_strategy_span_search(
+        input_ids: torch.tensor,
+        tokenizer, 
+        asst_tag: str="<|start_of_role|>assistant<|end_of_role|>",
```
[the remark this answers is not in the document]
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:852]: yes this is specific to granite actually, I dont have a way to configure it now

File: open_instruct/dataset_transformation.py (hunk 2 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py:871]: this is when the tag shapes are specified wrongly and are of the wrong size

Comment (garrett361) [reply open_instruct/dataset_transformation.py:871]: we don't want to error in those cases?

**Review by divya-kumari32 — APPROVED** [review]

File: open_instruct/dataset_transformation.py (hunk 2 above)
Comment (fabianlim) [reply open_instruct/dataset_transformation.py]: right now is a pass through.. we can put a warning

**Comment by fabianlim** [comment]
@garrett361 @Swanand-Kadhe I addressed both your comments

File: tests/test_chat_templates.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    spans = []
+    start = False
+    for i in range(len(labels)):
+        if labels[i] == ignore_label and not start:
+            spans.append([i])
+            start = True
+        elif labels[i] != ignore_label and start:
+            spans[-1].append(i)
+            start = False
+
```
Comment (garrett361) [inline tests/test_chat_templates.py:103]: Thanks for adding tests!

Is this an independent test, though? Looks like repeated logic to me. Thought we'd have something like explicitly asserting that the elements of `labels` which should be masked, are actually masked. Might be misunderstanding the test, though.

**Comment by fabianlim** [comment]
@garrett361 feel free to merge if its ok

**Review by dangxuanhong — COMMENTED** [review]
Thanks for supporting both jinja and jinja2. LGTM.
